src/main/java/org/zalando/planb/provider/OIDCController.java

```diff
@@ -64,16 +64,21 @@
      */
     @RequestMapping(value = {"/oauth2/access_token", "/z/oauth2/access_token"}, method = RequestMethod.POST)
     @ResponseBody
-    OIDCCreateTokenResponse createToken(@RequestParam(value = "realm", required = true) String realmName,
+    OIDCCreateTokenResponse createToken(@RequestParam(value = "realm") Optional<String> realmNameParam,
                                         @RequestParam(value = "grant_type", required = true) String grantType,
                                         @RequestParam(value = "username", required = true) String username,
                                         @RequestParam(value = "password", required = true) String password,
                                         @RequestParam(value = "scope") Optional<String> scope,
-                                        @RequestHeader(name = "Authorization") Optional<String> authorization)
+                                        @RequestHeader(name = "Authorization") Optional<String> authorization,
+                                        @RequestHeader(name = "Host") Optional<String> hostHeader)
             throws RealmAuthenticationException, RealmAuthorizationException, JOSEException {
         final Metric metric = new Metric(metricRegistry);
         metric.start();
 
+        final String realmName = realmNameParam.orElseGet(() -> realms.findRealmNameInHost(hostHeader
```

Comment:
:+1: totally right! didn't think about the sequential flow and only concentrate on hostHeader.get(). In case of hostHeader  = null it will end with the first Exception.